common/grpc/protobuf-jackson/src/main/java/org/curioswitch/common/protobuf/json/MessageMarshaller.java

```diff
@@ -0,0 +1,388 @@
+   */
+  public <T extends Message> String writeValueAsString(T message) throws IOException {
+    checkNotNull(message, "message");
+    SegmentedStringWriter sw = new SegmentedStringWriter(jsonFactory._getBufferRecycler());
+    JsonGenerator gen = jsonFactory.createGenerator(sw);
+    writeValue(message, gen);
+    return sw.getAndClear();
+  }
+
+  /**
+   * Converts a {@link Message} into JSON, writing to the provided {@link OutputStream}. Does not
+   * close the {@link OutputStream}.
+   */
+  public <T extends Message> void writeValue(T message, OutputStream out) throws IOException {
+    checkNotNull(message, "message");
+    checkNotNull(out, "out");
+    JsonGenerator gen = jsonFactory.createGenerator(out);
+    writeValue(message, gen);
+  }
+
+  /**
+   * Converts a {@link Message} into a JSON, writing to the provided {@link JsonGenerator}.
+   *
+   * @throws InvalidProtocolBufferException if there are unknown Any types in the message.
+   */
+  public <T extends Message> void writeValue(T message, JsonGenerator gen) throws IOException {
+    checkNotNull(message, "message");
+    checkNotNull(gen, "gen");
+    // TypeSpecificMarshaller for T.prototype is TypeSpecificMarshaller<T>
+    @SuppressWarnings("unchecked")
+    TypeSpecificMarshaller<T> serializer =
+        (TypeSpecificMarshaller<T>) registry.findForPrototype(message.getDefaultInstanceForType());
+    if (prettyPrinter != null) {
+      gen.setPrettyPrinter(prettyPrinter);
+    }
+    try {
+      serializer.writeValue(message, gen);
+    } catch (InvalidProtocolBufferException e) {
+      throw e;
+    } catch (IOException e) {
+      throw new InvalidProtocolBufferException(e);
+    } finally {
+      gen.close();
+    }
+  }
+
+  /**
+   * A {@link Builder} of {@link MessageMarshaller}s, allows registering {@link Message} types to
+   * marshall and set options.
+   */
+  public static final class Builder {
+
+    private boolean includingDefaultValueFields;
+    private boolean preservingProtoFieldNames;
+    private boolean omittingInsignificantWhitespace;
+    private boolean ignoringUnknownFields;
+
+    private final List<Message> prototypes = new ArrayList<>();
+
+    /**
+     * Registers the type of the provided {@link Message} for use with the created {@link
+     * MessageMarshaller}. While any instance of the type to register can be used, this will
+     * commonly be called with {@code getDefaultInstance()} on the type to register.
+     *
+     * <p>The provided {@link Message} and all nested {@link Message} types reachable from this one
+     * will be registered and available for marshalling. For clarity, it's generally a good idea to
+     * explicitly register any {@link Message} that you will pass to methods of {@link
+     * MessageMarshaller} even if they are already registered as a nested {@link Message}.
+     */
+    public Builder register(Message prototype) {
```

Comment:
How about adding the overloaded version that accepts `Class<? extends Message>`?